cache/cache.go

```diff
@@ -13,3 +13,180 @@
 // limitations under the License.
 
 package cache
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+	"sync/atomic"
+
+	rpctypes "go.etcd.io/etcd/api/v3/v3rpc/rpctypes"
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+var ErrCompacted = rpctypes.ErrCompacted
+
+type Stream struct{ Ch <-chan *clientv3.Event }
+
+// Cache buffers a single etcd Watch for a given key‐prefix and fan‑outs local watchers.
+// It is created internally by ShardedCache; callers typically use ShardedCache instead.
+type Cache struct {
+	prefix             string // keys this shard is responsible for, may be "" (root)
+	cfg                Config
+	client             *clientv3.Client
+	history            History
+	demux              *demux
+	ctx                context.Context
+	cancel             context.CancelFunc
+	latestCompactedRev int64         // updated atomically
+	ready              chan struct{} // closed when initialLoad finishes
+}
+
+// NewWithPrefix builds a cache shard that watches only the requested prefix.
+// For the root cache pass "".
+func NewWithPrefix(ctx context.Context, client *clientv3.Client, prefix string, opts ...Option) (*Cache, error) {
+
+	cfg := defaultConfig()
+	for _, opt := range opts {
+		opt(&cfg)
+	}
+
+	if cfg.HistoryWindowSize <= 0 {
+		return nil, fmt.Errorf("invalid HistoryWindowSize %d (must be > 0)", cfg.HistoryWindowSize)
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+
+	// choose history implementation
+	ringBuffer := newRingBuffer(cfg.HistoryWindowSize)
+
+	cache := &Cache{
+		prefix:  prefix,
+		cfg:     cfg,
+		client:  client,
+		history: ringBuffer,
+		demux:   newDemux(ringBuffer, cfg.ResyncInterval),
+		ctx:     ctx,
+		cancel:  cancel,
+		ready:   make(chan struct{}),
+	}
+
+	go func() {
+		// initial catch up
+		if err := cache.initialLoad(); err != nil {
+			log.Printf("cache initial load failed for prefix %q: %v",
+				prefix, err)
+		}
+		close(cache.ready)
+
+		// long-running upstream watch
+		serveWatchEvents(
+			ctx, client, prefix,
+			cache.demux.broadcast,
+			cache.demux,
+			cache.history,
+			func(rev int64) { atomic.StoreInt64(&cache.latestCompactedRev, rev) },
+			cfg.InitialBackoff,
+			cfg.MaxBackoff,
+		)
+	}()
+
+	return cache, nil
+}
+
+// initialLoad ranges over the prefix and seeds the ring.
+func (c *Cache) initialLoad() error {
+	opts := []clientv3.OpOption{clientv3.WithPrefix()}
+	resp, err := c.client.Get(c.ctx, c.prefix, opts...)
```

Comment:
Like I said before this is totally incorrect to load state (result of Get) and treat it as watch (result of Watch).